### PR TITLE
[SPARK-50245][SQL][TESTS] Extended CollationSuite and added tests where SortMergeJoin is forced

### PR DESCRIPTION
**Pull request by vladanvasi-db**
### What changes were proposed in this pull request?

I propose extending existing tests in `CollationSuite` and add cases where `SortMergeJoin` is forced and tested for correctness and use of `CollationKey`.

### Why are the changes needed?

These changes are needed to properly test behavior of join with collated data when different configs are enabled.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

The change is a test itself.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by vladanvasi-db** [comment]
@MaxGekk @stefankandic can you please take a look at this PR?

**Review by MaxGekk — COMMENTED** [review]
also cc @uros-db

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 1]
```diff
@@ -1598,11 +1598,38 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
             case b: HashJoin => b.leftKeys.head
           }.head.isInstanceOf[CollationKey])
         }
+
+        // Disable broadcast join to force sort merge join.
+        withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1") {
```
Comment (stefankandic) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:1603]: would it maybe be easier to run the tests with default and -1 values of the conf, and then just assert that different joins are used based on the conf's value?

Comment (uros-db) [reply sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:1603]: how about
```
Seq("-1", "1").foreach(val =>
  withSQLConf(SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> val) {
    ...
```

**Review by MaxGekk — COMMENTED** [review]
@vladanvasi-db If your PR affects tests only, please, add the `[TESTS]` tag to PR's title.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 2]
```diff
@@ -1562,47 +1596,52 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
...
-          assert(collectFirst(queryPlan) {
-            case b: HashJoin => b.leftKeys.head
-          }.head.isInstanceOf[CollationKey])
-        } else {
-          assert(!collectFirst(queryPlan) {
-            case b: HashJoin => b.leftKeys.head
-          }.head.isInstanceOf[CollationKey])
+          if (isSortMergeForced) {
+            // Only if collation doesn't support binary equality, collation key should be injected.
+            if (!CollationFactory.fetchCollation(t.collation).isUtf8BinaryType) {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:1621]: BTW, the `RewriteCollationJoin` checks `supportsBinaryEquality` which is a more stronger condition:
```
supportsBinaryEquality = !supportsSpaceTrimming && isUtf8BinaryType
```
Could you allign the test to `RewriteCollationJoin`.

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 3]
```diff
@@ -1562,47 +1596,52 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
...
+              assert(queryPlan.toString().contains("collationkey"))
+            } else {
+              assert(!queryPlan.toString().contains("collationkey"))
+            }
+          }
+          else {
+            // Only if collation doesn't support binary equality, collation key should be injected.
+            if (!CollationFactory.fetchCollation(t.collation).isUtf8BinaryType) {
+              assert(collectFirst(queryPlan) {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:1630]: Can you use `find`?

File: sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala [hunk 4]
```diff
@@ -1795,43 +1825,40 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
...
+
+          // confirm that right kind of join is used.
+          checkRightTypeOfJoinUsed(queryPlan)
+
+          // Only if collation doesn't support binary equality, collation key should be injected.
+          if (!CollationFactory.fetchCollation(t.collation).supportsBinaryEquality) {
+            assert(queryPlan.toString().contains("collationkey"))
+          } else {
+            assert(!queryPlan.toString().contains("collationkey"))
+          }
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala:1858]: Could you place the common code to a helper function.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @vladanvasi-db and @stefankandic @uros-db for review.

**Pull request closed without merging**